Add config entry for unsafe retries on container runner [ONPREM-1081] (#8778)
circleci/circleci-docs: jekyll/_cci2/container-runner.adoc

          - name: "other"
```

[#unsafe-retries]
=== Unsafe retries

> Unsafe retries enable container runner to automatically rerun tasks if there are unexpected interruptions during their execution. These disruptions could be due to network connectivity issues, the underlying node shutting down, or other unpredictable causes. This feature can also be used when scheduling workloads on spot instances, which often come with cost-saving benefits at the risk of pod preemptions with many Kubernetes providers.

Reviewer: Worth noting that errors surfaced in our UI as infrastructure failures should be expected to trigger an unsafe retry

> Unsafe retries is a feature that enables container runner to automatically rerun tasks if there are unexpected interruptions during their execution. These disruptions could be due to network connectivity issues, the underlying node shutting down, or other unpredictable causes. This feature can also be used when scheduling workloads on spot instances, which often come with cost-saving benefits at the risk of pod preemptions with many Kubernetes providers.

Reviewer 2: The singular/plural start of the paragraph feels a little awkward

[CAUTION]
====

> This feature is called “unsafe retries” for a reason. Unlike automatic retries on startup, retrying tasks during runtime can be risky. This is because tasks can have arbitrary steps that produce side effects which are not idempotent or stateless. It is important to use unsafe retries knowing the impact it may have on jobs and workflows. Use this feature with jobs and workflows that are fault-tolerant.

Reviewer: Instead of noting the impact it will have on jobs and workflows I think we would do well to warn about knowing the impact reran steps may have on external environments and resources such as production environments or databases. 

I'd also change "Use this feature with jobs and workflows that are fault tolerant" to "Use this feature with care, knowing the risks of rerunning jobs and workflows that may or may not be idempotent"

Author: This was lifted from Product wording/warning on the feature when previously presented to customers, but I can reword this to that.

====

Here is how unsafe retries work with container runner:

> 1. If a pod fails or gets evicted during runtime, container runner will "un-claim", or release the task.

Reviewer: un-claim may be exposing an implementation detail, I think we're okay to just say "release the task"

Author: Disagree. We talk about claiming and tasks in the docs [here](https://circleci.com/docs/container-runner/#resource-class-configuration-custom-pod). I think mentioning unclaiming here emphasizes that it may be claimed again by another instance of container agent.

> 2. All associated resources of the task are cleaned up.

Reviewer: s/All associated resources/All CircleCI Self Hosted Runner cluster resources/

This could give the impression that we'd somehow cleanup generated data or state external to the task management.

3. The released task then becomes available for reclaim by any container runner instance configured for the same resource class.

> 4. Any previously ran steps will be reset, and the task will restart from scratch.

Reviewer: The wording of "reset" here may also give the impression that we're undoing side effects somehow. Maybe simplify to "The task will be started from scratch, including previously ran steps".

I'm not sure about the usage of "task" vs "job" here. We don't really expose the concept of a task to customers

Author: We do expose the concept of tasks to customers: https://circleci.com/docs/faq/#what-is-a-CircleCI-task-vs-a-job

Also within this same document: https://circleci.com/docs/container-runner/#resource-class-configuration-custom-pod

          - name: "other"
```

[#unsafe-retries]
=== Unsafe retries

> Unsafe retries enable container runner to automatically rerun tasks that are unexpectedly interrupted during their execution. These disruptions could be due to network connectivity issues, the underlying node shutting down, or other unpredictable causes. Any job failure that would normally be displayed in the CircleCI web app as an infrastructure fail should be expected to trigger an unsafe retry when enabled.

Reviewer 3: ```suggestion
Unsafe retries enable container runner to automatically rerun tasks that are unexpectedly interrupted during their execution. These disruptions could be due to network connectivity issues, the underlying node shutting down, or other unpredictable causes. Any job failure that would be displayed in the CircleCI web app as an infrastructure fail should be expected to trigger an unsafe retry when enabled.
```

> This feature is also useful when scheduling workloads on spot instances, which often come with cost-saving benefits at the risk of pod preemptions with many Kubernetes providers.

Reviewer 3: ```suggestion
Unsafe retries is useful when scheduling workloads on spot instances, which often come with cost-saving benefits at the risk of pod preemptions with many Kubernetes providers.
```

[CAUTION]
====
This feature is called “unsafe retries” for a reason. Unlike automatic retries on startup, retrying tasks during runtime can be risky. This is because tasks can have arbitrary steps that produce external side effects which are not idempotent or stateless. This includes steps that could impact production environments or databases. Use this feature with care, knowing the risks of rerunning jobs and workflows that may or may not be idempotent.
====

> Here is how unsafe retries work with container runner:

Reviewer 3: ```suggestion
The following sequence shows how unsafe retires work:
```

1. If a pod fails or gets evicted during runtime, container runner will release the task.
2. All resources managed by container runner for the task, such as the Kubernetes pod and secret, are cleaned up and deleted.
3. The released task then becomes available for reclaim by any container runner instance configured for the same resource class.
4. Once reclaimed, the task is restarted completely from scratch, including previously run steps.
5. A task can be retried up to 3 times before it is deemed to have permanently failed.

> To enable unsafe retries, you need to set the `enableUnsafeRetries` flag in the <<resource-class-configuration-custom-pod,resource class configuration>> for each resource class. Here is an example where two resource classes are defined - one with unsafe retries enabled for spot instances and the other one where it is not enabled:

Reviewer 3: ```suggestion
To enable unsafe retries, set the `enableUnsafeRetries` flag in the <<resource-class-configuration-custom-pod,resource class configuration>> for each resource class. The following example shows two resource class definitions. Unsafe retries is enabled for the first, for spot instances, but not for the second resource class:
```
